### ISSUE TITLE
Feature request: Build ARM/ARM64 images

### ISSUE THREAD

**Issue by wxdao** [kiali/kiali#1091]
It would be great to have kiali/kiali on docker hub to work on ARM machines.

**Comment by pilhuhn** [comment]
Could you help us getting that done? 
Do you have arm hardware to test? I know AWS has Arm virtual machines, but setting all that up is not trivial

**Comment by wxdao** [comment]
I know shippable.com claims to provide free ARM/ARM64 machines for public repos. However I haven't had a chance to try their service yet.

I'd like to help you with this if possible. I just filed a contact to shippable.com to request free ARM machines, waiting for their response now.

**Comment by pilhuhn** [comment]
That's cool - thanks!

```
$GOACH=arm64
$GOOS=linux
```

[This post](https://medium.com/@mauridb/docker-multi-architecture-images-365a44c26be6) nicely explains how multi-arch images could then work. But I guess for the start we should only try to have the arm one available for people to explicitly select.

**Comment by wxdao** [comment]
Hi @pilhuhn. I managed to setup an arm pipeline using drone cloud. You can have a look at https://cloud.drone.io/wxdao/kiali/8 & https://github.com/wxdao/kiali.

Should I create a PR into the project if that looks right to you?

**Comment by israel-hdez** [comment]
Personally, I was expecting doing cross compilation using our current setup.
Currently, the docker images are build in a Jenkins box, so not sure how we could integrate drone into the build process.

But if this is OK for @pilhuhn, I guess you can create a PR.

**Comment by wxdao** [comment]
> Personally, I was expecting doing cross compilation using our current setup.
> Currently, the docker images are built in a Jenkins box, so not sure how we could integrate drone into the build process.
> 
> But if this is OK for @pilhuhn, I guess you can create a PR.

I agree that cross compilation is enough to build ARM docker images. I think what @pilhuhn wanted is to actually run tests in ARM machines before publishing ARM docker images, so you may not need to integrate drone to your internal jenkins box. Just checking the tests results from drone linked to this public repo before triggering a internal build should be enough.

**Comment by israel-hdez** [comment]
@wxdao Ok, so I understood a different thing
What I understood is to have an arm machine to test that the images are built correctly... A machine to download the arm images from dockerhub and try them (do a "smoke test").

So, let's wait for him to clarify ;)

**Comment by israel-hdez** [comment]
@wxdao are you still interested in this?

**Comment by wxdao** [comment]
Yes I am. However I'm not sure how I can help.

**Comment by israel-hdez** [comment]
Perhaps, as was briefly discussed before, just finish the CI setup, so that tests are run on each PR. So, you could just finish the initial effort and open a PR.
Once that is done, we can check how to cross-compile and publish the images to DockerHub.

**Comment by cfcosta** [comment]
@wxdao did you test running Istio/Grafana/Prometheus/Jaeger on ARM? Because we currently depend on those. There should be no reason for it not to work on ARM, but if everything else is not working, then there's no point.

Also, testing it on k3s would be amazing.

**Comment by lucasponce** [comment]
This issue has no activity for months.
It sounds interesting as a community effort, so, if someone would like to help, feel free to ping us and we can re-open it.

**Comment by MrXinWang** [comment]
@lucasponce @wxdao @israel-hdez Hi all, is this arm64 supporting stuff still on going? If it is, can I help in pushing things forward?

**Comment by MrXinWang** [comment]
So here is my initial plan:

I will firstly try to build binary and images on my arm64 machine to make it buildable. If  successful I will try to post PR to fix this first.

Then I will try to discuss with the community in order to update the manifest to make the image multiarch.

@israel-hdez @pilhuhn @lucasponce Wondering if the community has any comments (highly welcomed!) Thanks!

**Comment by israel-hdez** [comment]
@MrXinWang If you are willing to help, we can re-open this issue and assign it to you.

Our main concern was how to run the unit tests on CI for every PR to ensure that builds won't get broken. However, I just saw this post in Travis's blog which may make this easier: https://blog.travis-ci.com/2019-10-07-multi-cpu-architecture-support

Our second concern was how to build and push the multi-arch images to Quay.io and DockerHub. The box building the releases doesn't have access to ARM hardware, so we would need to do cross-compiling and so.

We are not sure how hard this is to do. But if you are willing to help, the team will he happy to assist you and bring guidance :-)

**Comment by MrXinWang** [comment]
Hi @israel-hdez ! Thanks for your reply and the team‘s kindly assistance. Discussed with my manager and yes we would love to work on this issue :)))

Also completely agree with the CI stuff. I think not only we need to keep the consistancy of the current CI, also we need to possibly add CI for arm to test its functionality on arm.

About the multiarch image, I dont know how the community build the image now, but it should be not that difficult, as <1> if the community has arm64 machine (either a server or a raspberry pi 4), the image can be directly build (I need to propose some PRs first); <2> if the community only has x86_64 machines, we can also build arm64 image by using `docker buildx`.

What do you think? Maybe I can propose a PR first to fix the Dockerfile and Makefile on arm64 first? Any other suggestions?

**Comment by israel-hdez** [comment]
@MrXinWang Thanks for taking this issue!
I reopened and assigned it to you.

Doing the changes to Dockerfile and Makefile to have a container image for arm64 sounds a good first step. Once you have that, I guess it shouldn't be too hard to have Travis running the unit tests for arm64 (I guess this could be a good 2nd step). So, go ahead with the PR :-)

Then, perhaps adding some instructions on README.md to build Kiali on arm64 would be great. This will ensure that knowledge is shared.

Once there is a way to manually build the arm64 container and CI in place, we can start looking at how to push the arm64 container do DockerHub & Quay.io.

**Comment by MrXinWang** [comment]
@israel-hdez PR #2532 proposed. First step done, although I am not sure if the operator image should also be enabled...

For the operator image, do you suggest we ask the open-operator-sdk to propose arm releases or we build the binary from source?

As suggested, I am then going to try travis. Any comments do let me know please :)))

**Comment by israel-hdez** [comment]
> For the operator image, if it is required, do you suggest we ask the open-operator-sdk to propose arm releases or we build the binary from source?

The operator is the preferred way to install the latest Kiali, so I think it's better to ask. But it may take time until they add ARM64 support. In the meanwhile, building from source would be quite fine because I think we don't want the installation experience to be different between x86_64 vs arm. Updating two different install instructions is cumbersome.

Otherwise, if we skip the operator, how do you think people will install Kiali on ARM?

**Comment by MrXinWang** [comment]
Hi @israel-hdez ! Thanks for letting me know. I have proposed [operator-sdk/issues#2686](https://github.com/operator-framework/operator-sdk/issues/2686), can you please also leave some comments there to push this process? If they release that it will be super convenient for us to enable the arm64 stuff.

Also I saw the comments you left in the PR, if we really need the operator image, the conditional used currently can be removed. I will firstly work on the source code building.

**Comment by israel-hdez** [comment]
Re-opening, because there is still work to do :)

**Comment by jmazzitelli** [comment]
> Re-opening, because there is still work to do :)

What is the remaining work? Is it the Kiali Operator build?

**Comment by MrXinWang** [comment]
@israel-hdez @jmazzitelli Yep, the main part is the operator image, or the operator-sdk arm64 binary in a way...

Also I think we need to add travis support for arm, as @israel-hdez mentioned here https://blog.travis-ci.com/2019-10-07-multi-cpu-architecture-support

**Comment by BlackMix** [comment]
Good job 👏👏
I liked this system and wanted to use it on my 
raspberry pi 4b 4gb, I'm looking forward to testing it

**Comment by israel-hdez** [comment]
I see progress for the operator-sdk support for arm64 :smiley: 
Nice to see this is progressing.

Thanks for this @BlackMix @MrXinWang

**Comment by BlackMix** [comment]
@MrXinWang can u test? [operator-sdk-arm64](https://github.com/BlackMix/operator-sdk-arm64/releases/tag/v0.16.0)
I did build and posted, let me know

**Comment by MrXinWang** [comment]
@BlackMix Perfect! Awesome work!

```shell
root@entos-ampere:~# uname -a
Linux <hostname> 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:41:03 UTC 2018 aarch64 aarch64 aarch64 GNU/Linux

root@<hostname>:~# ./operator-sdk-v0.16.0-27-g3687cc1a-dirty-aarch64-linux-gnu
An SDK for building operators with ease

Usage:
  operator-sdk [command]

Available Commands:
  add             Adds a controller or resource to the project
  build           Compiles code and builds artifacts
  bundle          Work with operator bundle metadata and bundle images
  cleanup         Delete and clean up after a running Operator
  completion      Generators for shell completions
  generate        Invokes a specific generator
  help            Help about any command
  migrate         Adds source code to an operator
  new             Creates a new operator application
  olm             Manage the Operator Lifecycle Manager installation in your cluster
  print-deps      Print Golang packages and versions required to run the operator
  run             Run an Operator in a variety of environments
  scorecard       Run scorecard tests
  test            Tests the operator
  version         Prints the version of operator-sdk

Flags:
  -h, --help      help for operator-sdk
      --verbose   Enable verbose logging

Use "operator-sdk [command] --help" for more information about a command.
```
@israel-hdez And now I can propose another PR to enable the operator image on arm. However, I guess we have another choice to make: As @BlackMix proposed his patch already to the operator-sdk community, shall we wait for the operator-sdk official release or we use @BlackMix's release in his own repository in the building script? Any comments?

**Comment by BlackMix** [comment]
so kiali need istio right? is istio for arm64? or I can use without istio, never tesed it first time xD

**Comment by MrXinWang** [comment]
@BlackMix Some of my colleagues are working on that, should be out soon :))

**Comment by BlackMix** [comment]
so I'll wait everyone official finish it,
kiali and operator are build 100% success and working for me tho.
I just need istio now :)

**Comment by jmazzitelli** [comment]
> As @BlackMix proposed his patch already to the operator-sdk community, shall we wait for the operator-sdk official release or we use @BlackMix's release in his own repository in the building script

@MrXinWang you can create a draft PR to get started on it. But we will not merge into master anything using a personal repo (such as @BlackMix )... but get the draft PR working and when operator-sdk releases an official ARM release, we can update the draft PR and then promote it and test/merge it.

But for sure you can feel free to start this work now and submit a draft PR for the kiali team to start looking at and reviewing.

**Comment by MrXinWang** [comment]
@jmazzitelli PR #2581 created as a draft (Well actually worse than a draft because I only tested the building can be successfully executed on both x86_64 and aarch64, and a huge lot of things need to add...).

**Comment by BlackMix** [comment]
[test build passed from operator](https://travis-ci.org/github/BlackMix/operator-sdk/builds/669385074) all builds success.
just waiting merger now.

**Comment by lucasponce** [comment]
cc @jmazzitelli to take a look and update/finish this one.

**Comment by jmazzitelli** [comment]
We have travis running ARM64 builds for both kiali and the operator:

https://github.com/kiali/kiali/blob/master/.travis.yml#L5
https://github.com/kiali/kiali-operator/blob/master/.travis.yml#L6

**Comment by jmazzitelli** [comment]
NOTE: the Kiali team is not prepared to support Kiali or the Kiali operator on ARM64 hardware. So while we can say it builds (Travis CI confirms this), we cannot say anything as to its functionality nor can we support it. So, right now, we are not shipping any images to Quay.io's [kiali](https://quay.io/repository/kiali/kiali?tab=tags) or [kiali-operator](https://quay.io/repository/kiali/kiali-operator?tab=tags) repositories. People can build them locally on their own ARM64 hardware using Makefiles if they want to test it out.

**Comment by israel-hdez** [comment]
We can use [docker's buildx](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) and [golang cross-compiling](https://dave.cheney.net/2015/08/22/cross-compilation-with-go-1-5) to build the arm64 container image. But docker's buildx is an experimental feature. Not sure if we will be able to have it in the Jenkins server.

**Comment by lucasponce** [comment]
It's not clear what should be the action from here.
To make travis to push these arch as best effort ?

**Comment by jmazzitelli** [comment]
All the work is done EXCEPT we are pushing arm64 binaries to quay.io. We would have to do something in travis.

If you want arm64 support, you can build the binaries locally on your arm64 machine. We do not do anything with cross-compiling.

**Comment by israel-hdez** [comment]
Moving out of the sprint, because this seems to be in a stall.

**Comment by morlay** [comment]
I have build oci image for 1.20.0 https://hub.docker.com/r/querycap/kiali/tags by https://github.com/querycap/kiali/blob/master/Dockerfile

but have some issue for pr fix this issue.

@israel-hdez a question.

why amd64 use `registry.access.redhat.com/ubi7-minimal` arm64 use `registry.access.redhat.com/ubi8-minimal`

https://github.com/kiali/kiali/tree/master/deploy/docker

**Comment by jmazzitelli** [comment]
> why amd64 use registry.access.redhat.com/ubi7-minimal arm64 use registry.access.redhat.com/ubi8-minimal

Because only UBI 8 supports arm64. Containers for arm64 are not supported on RHEL 7 or UBI 7.

See https://bugzilla.redhat.com/show_bug.cgi?id=1728771

**Comment by morlay** [comment]
@jmazzitelli got it.

but why not upgrade to ubi8 both for arm64 & amd64 ?

**Comment by jmazzitelli** [comment]
> but why not upgrade to ubi8 both for arm64 & amd64 ?

Most UBI8 images will work on a RHEL7 bare metal host, however those images are not officially supported by RHEL. So if there is a problem the responsibility falls on the container owner to work it out.

This means we currently are not guaranteed that an image based on UBI8 will work 100% on RHEL7 machines (and we do not want to break those users whose bare metal machines are RHEL7).

**Comment by morlay** [comment]
@jmazzitelli Thanks.

for now, have to build multi platforms image unofficially

**Comment by morlay** [comment]
@jmazzitelli could you check this PR https://github.com/kiali/kiali/pull/3405

I find a way to avoid change base image of amd64, but could add arm64 image support.

If no others concern, we could setup buildx in CI, and publish multi-arch images

**Comment by israel-hdez** [comment]
I think there are no remaining tasks for this issue. Looks completed as we have been publishing container images for amd64, arm64, s390x and ppc64le since about version 1.28 of Kiali.

Closing as done.
